File: back/MarsRoversApi/src/main/kotlin/rover/Robot.kt
package rover

class Robot(x: Int, y: Int, direction: String) {

    private var direction: Direction = Direction.of(direction)
    private var position : Position = Position(x, y)

    fun move(vararg movements: String) {
        movements.forEach { movement ->
            move(movement)
        }
    }

    fun isInPosition(x:Int, y: Int): Boolean {
        return this.position == Position(x, y)
    }

    private fun forward() {
        position = direction.forward(position)
    }

    private fun backwards() {
        position = direction.backward(position)
    }

    private fun rotate(direction: String) {
        this.direction = Direction.of(direction)
    }

    private fun move (movement: String){
        if(movement == "F") {
            forward()
            return
        }
        if(movement == "B") {
            backwards()
            return
        }
        rotate(movement)
    }
}

File: back/src/MarsRover/Domain/Robot.cs
using MarsRover.Domain.Orientations;
using MarsRover.Monad;

namespace MarsRover.Domain
{
    public class Robot
    {
        private Position position;
        private Orientation orientation;

        public Robot(Map map, int x, int y, string orientation)
        {
            position = new Position(map, x, y);
            this.orientation = OrientationFactory.Create(orientation);
        }

        public Situation GetSituation()
        {
            return new Situation(position.x, position.y, orientation.ToString());
        }

        public Either<Error, Robot> Move(string movements)
        {
            var result = Either<Error, Robot>.Success(this);
            foreach (var movement in movements)
            {
                result = result.Bind(_ => Move(movement));
            }
            return result;
        }

        private Either<Error, Robot> Move(char movement)
        {
            if(movement == 'F')
            {
                return Forward();

            }
            if (movement == 'B')
            {
                return Backward();
            }

            orientation = OrientationFactory.Create(movement.ToString());
            return Either<Error, Robot>.Success(this);
        }

        private Either<Error, Robot> Forward()
        {
            return orientation.Forward(position)
                .Bind(UpdatePosition)
                .Match(
                    onSuccess: _ => Either<Error, Robot>.Success(this),
                    onError: Either<Error, Robot>.Error
            );
        }

        private Either<Error, Robot> Backward()
        {
            return orientation.Backward(position)
                .Bind(UpdatePosition)
                .Match(
                    onSuccess: _ => Either<Error, Robot>.Success(this),
                    onError: Either<Error, Robot>.Error
                );
        }

        private Either<Error, Position> UpdatePosition(Position position)
        {
            this.position = position;
            return Either<Error, Position>.Success(position);
        }
    }
}
